src/Psl/IO/Internal/ResourceHandle.php

```diff
@@ -0,0 +1,189 @@
+     * @throws Psl\IO\Exception\BlockingException       If the handle is a socket or similar, and the write would block.
+     * @throws Psl\IO\Exception\RuntimeException        If an error occurred during the operation.
+     */
+    public function write(string $bytes): int
+    {
+        return $this->box(function () use ($bytes) {
+            $metadata = stream_get_meta_data($this->resource);
+            if ($metadata['blocking']) {
+                throw new Psl\IO\Exception\BlockingException('The stream is blocking.');
+            }
+
+            $result = fwrite($this->resource, $bytes);
+            if ($result === false) {
+                $error = error_get_last();
+
+                throw new Psl\IO\Exception\RuntimeException($error['message']);
+            }
+
+            return $result;
+        });
+    }
+
+    /**
+     * @throws Psl\IO\Exception\AlreadyClosedException  If the handle has been already closed.
+     * @throws Psl\IO\Exception\RuntimeException        If an error occurred during the operation.
+     */
+    public function seek(int $offset): void
+    {
+        $this->box(function () use ($offset) {
+            $metadata = stream_get_meta_data($this->resource);
+            Psl\invariant($metadata['seekable'], 'Stream is not seekable.');
+
+            $result = fseek($this->resource, $offset);
+            if (0 !== $result) {
+                throw new Psl\IO\Exception\RuntimeException('Failed to seek the specified position.');
+            }
+        });
+    }
+
+    /**
+     * @throws Psl\IO\Exception\AlreadyClosedException  If the handle has been already closed.
+     * @throws Psl\IO\Exception\RuntimeException        If an error occurred during the operation.
+     */
+    public function tell(): int
+    {
+        return $this->box(function () {
+            $metadata = stream_get_meta_data($this->resource);
+            Psl\invariant($metadata['seekable'], 'Stream is not seekable.');
+
+            $result = ftell($this->resource);
+            if ($result === false) {
+                $error = error_get_last();
+
+                throw new Psl\IO\Exception\RuntimeException($error['message']);
+            }
+
+            return $result;
+        });
+    }
+
+    /**
+     * @throws Psl\IO\Exception\AlreadyClosedException  If the handle has been already closed.
+     * @throws Psl\IO\Exception\BlockingException       If the handle is a socket or similar, and the read would block.
+     * @throws Psl\IO\Exception\RuntimeException        If an error occurred during the operation.
+     * @throws InvariantViolationException              If $max_bytes is 0.
+     */
+    public function read(?int $max_bytes = null): string
+    {
+        return $this->box(function () use ($max_bytes) {
+            Psl\invariant($max_bytes === null || $max_bytes > 0, '$max_bytes must be null, or > 0');
+            $metadata = stream_get_meta_data($this->resource);
+            if ($metadata['blocking']) {
```

Comment:
How does non-blocking functionality behaves exactly? 
In non-blocking frameworks like amphp, you have a promise or coroutine  for those kind of actions.
Now you return similar results as the regular f* resource functions.

Does the non-blocking part matters much here?
Or is it just writing to the resource whilst running the next lines of code?

Reply:
blocking mode ensures that the stream returns right away when calling `fread` for example, instead of waiting for data to be available ( e.g: socket ). ( see https://www.php.net/manual/en/function.stream-set-blocking.php )

Reply:
Alrighty, makes sense then. 
The documentation on php.net is quite confusing though :)